snow/engine/snowman/transitive.go

```diff
@@ -1103,7 +1114,59 @@ func (t *Transitive) addUnverifiedBlockToConsensus(
 	)
 	return true, t.Consensus.Add(ctx, &memoryBlock{
 		Block:   blk,
-		metrics: &t.metrics,
+		metrics: t.metrics,
 		tree:    t.nonVerifieds,
 	})
 }
+
+// getProcessingAncestor finds [initialVote]'s most recent ancestor that is
+// processing in consensus. If no ancestor could be found, false is returned.
+//
+// Note: If [initialVote] is processing, then [initialVote] will be returned.
+func (t *Transitive) getProcessingAncestor(ctx context.Context, initialVote ids.ID) (ids.ID, bool) {
+	// If [bubbledVote] != [initialVote], it is guaranteed that [bubbledVote] is
+	// in processing. Otherwise, we attempt to iterate through any blocks we
+	// have at our disposal as a best-effort mechanism to find a valid ancestor.
+	bubbledVote := t.nonVerifieds.GetAncestor(initialVote)
+	for {
+		if t.Consensus.Processing(bubbledVote) {
+			t.Ctx.Log.Verbo("applying vote",
+				zap.Stringer("initialVoteID", initialVote),
+				zap.Stringer("bubbledVoteID", bubbledVote),
+			)
+			if bubbledVote != initialVote {
+				t.numProcessingAncestorFetchesSucceeded.Inc()
+			} else {
+				t.numProcessingAncestorFetchesUnneeded.Inc()
+			}
+			return bubbledVote, true
+		}
```

Comment:
Moving the `Processing` check before `getBlock` avoids calling `VM.GetBlock` when `getProcessingAncestor` is called with a processing block.